ddtrace/contrib/anthropic/_streaming.py

```diff
@@ -0,0 +1,322 @@
+            stream,
+            self._dd_integration,
+            self._dd_span,
+            self._args,
+            self._kwargs,
+        )
+        traced_stream.text_stream = traced_stream._text_stream()
+        return traced_stream
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.__wrapped__.__aexit__(exc_type, exc_val, exc_tb)
+
+
+def _process_finished_stream(integration, span, args, kwargs, streamed_chunks):
+    # builds the response message given streamed chunks and sets according span tags
+    resp_message = {}
+    try:
+        resp_message = _construct_message(streamed_chunks)
+
+        if integration.is_pc_sampled_span(span):
+            _tag_streamed_chat_completion_response(integration, span, resp_message)
+        if integration.is_pc_sampled_llmobs(span):
+            integration.llmobs_set_tags(
+                span=span,
+                resp=resp_message,
+                args=args,
+                kwargs=kwargs,
+            )
+    except Exception:
+        log.warning("Error processing streamed completion/chat response.", exc_info=True)
+
+
+def _construct_message(streamed_chunks):
+    """Iteratively build up a response message from streamed chunks.
+
+    The resulting message dictionary is of form:
+      {"content": [{"type": [TYPE], "text": "[TEXT]"}], "role": "...", "finish_reason": "...", "usage": ...}
+    """
+    message = {"content": []}
+    for chunk in streamed_chunks:
+        message = _extract_from_chunk(chunk, message)
+
+        if "finish_reason" in message:
+            return message
+    return message
+
+
+def _extract_from_chunk(chunk, message={}) -> Tuple[Dict[str, str], bool]:
+    """Constructs a chat message dictionary from streamed chunks given chunk type"""
+    TRANSFORMATIONS_BY_BLOCK_TYPE = {
+        "message_start": _on_message_start_chunk,
+        "content_block_start": _on_content_block_start_chunk,
+        "content_block_delta": _on_content_block_delta_chunk,
+        "message_delta": _on_message_delta_chunk,
+    }
+    chunk_type = getattr(chunk, "type", "")
+    transformation = TRANSFORMATIONS_BY_BLOCK_TYPE.get(chunk_type)
+    if transformation is not None:
+        message = transformation(chunk, message)
+
+    return message
+
+
+def _on_message_start_chunk(chunk, message):
+    # this is the starting chunk of the message
+    if getattr(chunk, "type", "") != "message_start":
+        return message
+
+    chunk_message = getattr(chunk, "message", "")
+    if chunk_message:
+        content_text = ""
+        contents = getattr(chunk.message, "content", [])
+        for content in contents:
+            if content.type == "text":
+                content_text += content.text
+                content_type = "text"
+            elif content.type == "image":
+                content_text = "([IMAGE DETECTED])"
+                content_type = "image"
+            message["content"].append({"text": content_text, "type": content_type})
```

Comment:
https://github.com/DataDog/dd-trace-py/pull/9420/files#r1624529869 Can you confirm whether or not we need to process content from message_start content blocks? According to their docs (and local testing) message_start content blocks are always empty.

Reply:
Have you tested with the `.stream()` function instead of `.create()`? That function builds the whole message into the first content block.

Reply:
here's an example of a fully built up message start event:. This is the result of the stream being fully iterated through, and the stream manager building up the initial event to contain all content and tokens. Also note that our output_tokens is 15, leading me to believe we don't need that initial 1 output token from the message start event (for basic streams by calling `.create(stream=true)`, where the start event will not "built up" like below


`MessageStartEvent(message=Message(id='msg_017z3e6QB2VQhUBqF9zuLmiK', content=[TextBlock(text='The famous philosophical statement "I think, therefore I am" (originally in', type='text')], model='claude-3-opus-20240229', role='assistant', stop_reason='max_tokens', stop_sequence=None, type='message', usage=Usage(input_tokens=27, output_tokens=15)), type='message_start')`

Reply:
Resolved offline - we'll skip trying to process the fully built up message (via start/end events) as this will likely break if anthropic introduces multi chat outputs.